### PR TITLE
Adding azure pipeline for nimi-bot pytests

### PR DESCRIPTION
**Pull request by SukruthGrandhiNI**
- [ ] ~~This contribution adheres to [CONTRIBUTING.md](https://github.com/ni/nimi-python/blob/master/CONTRIBUTING.md).~~

- [ ] ~~I've updated [CHANGELOG.md](https://github.com/ni/nimi-python/blob/master/CHANGELOG.md) if applicable.~~

- [ ] ~~I've added tests applicable for this pull request~~

### What does this Pull Request accomplish?

In this Pull Request added a functionality to run system tests using azure pipelines.
This azure pipeline will be run for every new PR raised in this repository.
Example- https://ni.visualstudio.com/Users/_build/results?buildId=1312088&view=results

Note: we are using Azure pipelines instead of github actions for the security reasons. In future we will change to github actions after the vms are created on the AWS cloud(Non NI network). Security constraints are captured as comments in https://ni.visualstudio.com/DevCentral/_sprints/backlog/MTO%20NIB/DevCentral/21Q3/09?workitem=1534568

Feature specifications.
1. Total 9 modules system tests are run in the pipeline. Please refer the above link for the list of modules.
2. Systems tests for each OS will run parallelly using staging. Also, each module system tests are ran parallel in a given OS using Jobs. 
2.  Each module system tests are run for python 36,37,38 and 39 versions.
3.  Each operating system are defined as a separate stage in the pipeline, this PR has changes only for testing on windows 10 OS.
4.  To reduce pipeline build time, in a single machine there are 4 agents created. That means 4 modules system tests can run parallel.
5. Module system tests are run only after validating the codegen.
6. Added new agent pool called "nimi-bot"(https://ni.visualstudio.com/_settings/agentpools?poolId=474&view=agents), all the agents related to nimi-bot are hosted here.
7. Created two stages, first stage will verify codegen is up-to date, in the second stage we are testing system tests for win64.

Below tasks are addressed after submitting the changes
1. Need to create a official azure pipeline which is linked to master branch of nimi-bot.
2. If the PR is raised by external contributor then pipeline build should start only after approval from the owner of this repo. This is done for the sake of security.
3. Adding new stage for Linux OS - OS Flavor To be defined.(Expected within this iteration).
4. Adding new stage for win32(python 32).
5. Currently agent machines in NI domain are created manually- this need to automated using IaC(Expectation for windows agents within this iteration).
6. Adding the pipeline build status badge to the readme file.

Pipeline log: https://ni.visualstudio.com/Users/_build/results?buildId=1364419&view=logs&j=82de4692-693a-5bea-5cad-e610b6a98a8a&t=82447a30-d41f-508c-a457-57990d8dcc66

Future scope
1. Creation of agents in Non NI network, these agents machines will be hosted in AWS, these agents are also created using IaC.
2. Change Azure pipelines to Github actions - so that external contributors can view the logs.

Few common issues
1) Due to system tests running parallelly across different python versions are trying to access the common shared resources.
   Example1 - https://ni.visualstudio.com/Users/_build/results?buildId=1312059&view=logs&j=d35c6cb8-fa2f-53f1-119d-cb1f7fc8a3fc&t=4eb9ba65-1bf8-5117-d969-a06c9958fabb
  Example2 - https://ni.visualstudio.com/8c36cb1d-ece7-4ec9-9c6b-409d081af0e8/_apis/build/builds/1310877/logs/17
Above issues are not reproducible after changing parallel variable to 1 in system_tests.bat.

### What testing has been done?

Created a pipeline for fork branch(https://github.com/SukruthGrandhiNI/nimi-python/),
Test PR pipeline trigger on the fork branch by raising the PR: https://github.com/SukruthGrandhiNI/nimi-python/pull/2

### CODE REVIEW

File: azure-pipelines.yml [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+              os.system("tox -e codegen")
+              os.system('git commit -a -m "test"')
+              commit_id_after_code_generator = os.system("git log -1 --format=%h")
+              if commit_id_before_code_generator != commit_id_after_code_generator:
+                print("Please re-generate the codegen using command 'tox -e codegen'", file=sys.stderr)
+                sys.exit(-1)
+
+              # To verify code generated is not added in .gitignore list
+              output = subprocess.check_output(["git", "ls-files", "--others", "-i", "--exclude-standard"], shell=True).decode()
+              files_ignored = [ file_path for file_path in output.split("\n") if not file_path.startswith((".tox", 'commands.log', 'generated/driver_file_built', 'generated/printed_global_files', 'tools/__pycache__')) and file_path]
```
Comment (SukruthGrandhiNI) [inline azure-pipelines.yml:40]: Added ".tox", 'commands.log', 'generated/driver_file_built', 'generated/printed_global_files', 'tools/__pycache__' in the exception list
https://ni.visualstudio.com/Users/_build/results?buildId=1312033&view=logs&j=45c68899-b372-5281-6fc3-3df8e4c59a4a&t=7ed62bfa-b7ac-58d9-69f8-a6c96c9ca2af

**Comment by sbethur** [comment]
ok to test

File: azure-pipelines.yml [hunk 2]
```diff
@@ -0,0 +1,128 @@
+# A pipeline with no CI trigger
+trigger: none
+
+stages:
+  - stage: Windows_nimi_bot
```
Comment (sbethur) [inline azure-pipelines.yml:5]:
```suggestion
  - stage: nimi_bot_windows
```

File: azure-pipelines.yml [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+
+              # To verify code is generated
+              os.system('git config user.email "dummy@ni.com"')
+              os.system('git config user.name "dummy"')
+              commit_id_before_code_generator = os.system("git log -1 --format=%h")
+              os.system("tox -e codegen")
+              os.system('git commit -a -m "test"')
+              commit_id_after_code_generator = os.system("git log -1 --format=%h")
+              if commit_id_before_code_generator != commit_id_after_code_generator:
+                print("Please re-generate the codegen using command 'tox -e codegen'", file=sys.stderr)
```
Comment (sbethur) [inline azure-pipelines.yml:33]:
```suggestion
                print("Run codegen and include the generated files in the PR", file=sys.stderr)
```

File: azure-pipelines.yml [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+              # To verify code generated is not added in .gitignore list
+              output = subprocess.check_output(["git", "ls-files", "--others", "-i", "--exclude-standard"], shell=True).decode()
+              files_ignored = [ file_path for file_path in output.split("\n") if not file_path.startswith((".tox", 'commands.log', 'generated/driver_file_built', 'generated/printed_global_files', 'tools/__pycache__')) and file_path]
+              if files_ignored:
+                print(f"{files_ignored} are generated using 'tox -e codegen', these files are not submitted as part of PR change due to paths added in .gitignore", file=sys.stderr)
+                sys.exit(-1)
+
+              sys.exit(0)
+
+    - job: Windows_Running_nidcpower_system_tests
```
Comment (sbethur) [inline azure-pipelines.yml:45]:
```suggestion
    - job: nidcpower_system_tests_windows
```

File: azure-pipelines.yml [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+      dependsOn: Windows_nimi_bot_Validate_Codegen 
+      workspace:
+        clean: all
+      steps:
+        - script: | 
+            tools\system_tests.bat nidcpower
+          displayName: Running nidcpower system tests
```
Comment (sbethur) [inline azure-pipelines.yml:52]:
```suggestion
          displayName: nidcpower system tests 
```

**Review by sbethur — CHANGES_REQUESTED** [review]

**Comment by sbethur** [comment]
test this please

**Comment by sbethur** [comment]
test this please

File: azure-pipelines.yml (hunk 2 above)
Comment (SukruthGrandhiNI) [reply azure-pipelines.yml]: done

File: azure-pipelines.yml (hunk 5 above)
Comment (SukruthGrandhiNI) [reply azure-pipelines.yml:76]: "nidcpower system tests windows"

**Comment by marcoskirsch** [comment]
Validate codegen step:

https://ni.visualstudio.com/Users/_build/results?buildId=1312088&view=logs&j=45c68899-b372-5281-6fc3-3df8e4c59a4a&t=7ed62bfa-b7ac-58d9-69f8-a6c96c9ca2af

Looks like a no-op. Perhaps we need to make clean first?

**Comment by marcoskirsch** [comment]
There's a lot of warnings in the tests eg
https://ni.visualstudio.com/Users/_build/results?buildId=1312088&view=logs&j=d085f2e7-0dd0-5ff9-f8e0-b873e7000a7a&t=4fe511f4-64d3-5abb-22fb-8f41a176708d&l=944

**Comment by marcoskirsch** [comment]
Some of the tests e.g. nitclk and nidigital took 8+ minutes.
Is this in line with the Jenkins based system? Seems like an eternity and would like to know more about it.

**Comment by marcoskirsch** [comment]
Right now we have the ability to run on 32-bit and 64-bit Windows OSes. This is actually important because we have bit specific code to deal with different packing in either platform.

In this AzDO flavor it isn't clear which bitness this is running on. Can you make clear the bitness? Can you comment on the other bitness and Linux plans?

**Comment by marcoskirsch** [comment]
I noticed agent names are "nimibot_windows"
We consistently spell "nimi-bot". If dash is not permitted, then underscore would match what you use in the yml in this review.

Also, instead of windows consider win32 and win64.

File: tools/system_tests.bat [hunk 6]
```diff
@@ -4,7 +4,7 @@ rem We don't need no spinner - mainly because we aren't running tox interactivel
 set TOX_PARALLEL_NO_SPINNER=1
 rem Run up to 3 environments in parallel, but still send the output to stdout
 rem We only run the tests if the tox file exists
-c:\python38\python.exe -m tox -c tox-system_tests.ini --parallel 3 --parallel-live
+python.exe -m tox -c tox-system_tests.ini --parallel 1 --parallel-live
```
Comment (sbethur) [inline tools/system_tests.bat:7]: The reason 3 didn't work is because sometimes the agents try to create `junit` and `kibana` folders simultaneously and fail. We got around this by creating the folders before running the tests. See https://github.com/ni/nimi-python/blob/e12ac5b7854f8bdfbfdbc4e21afe9a17fb524aa7/tools/system_test_jobs.groovy#L94

File: azure-pipelines.yml [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+              os.system('git config user.name "dummy"')
+              commit_id_before_code_generator = os.system("git log -1 --format=%h")
+              os.system("tox -e codegen")
+              os.system('git commit -a -m "test"')
+              commit_id_after_code_generator = os.system("git log -1 --format=%h")
+              if commit_id_before_code_generator != commit_id_after_code_generator:
+                print("Run codegen and include the generated files in the PR", file=sys.stderr)
+                sys.exit(-1)
+
+              # To verify code generated is not added in .gitignore list
```
Comment (sbethur) [inline azure-pipelines.yml:59]: You can get rid of this, it doesn't do what I was thinking of, which is check if the PR is creating new unintended git-ignored files. I don't think there's a good way to check that.

**Comment by SukruthGrandhiNI** [comment]
> Right now we have the ability to run on 32-bit and 64-bit Windows OSes. This is actually important because we have bit specific code to deal with different packing in either platform.
> 
> In this AzDO flavor it isn't clear which bitness this is running on. Can you make clear the bitness? Can you comment on the other bitness and Linux plans?

**Comment by SukruthGrandhiNI** [comment]
> Right now we have the ability to run on 32-bit and 64-bit Windows OSes. This is actually important because we have bit specific code to deal with different packing in either platform.

 32 windows is not in OS supported list, Please refer the link: https://dev.azure.com/ni/DevCentral/_git/Policy?path=%2Fproduct_and_technology%2FOperating-System-Support.md&version=GBmaster&_a=preview

> In this AzDO flavor it isn't clear which bitness this is running on. Can you make clear the bitness? Can you comment on the other bitness and Linux plans?
Since 32 bit windows is not there, so I am thinking to not to add bitness here. If you want to explicitly call out I can make the change.

File: azure-pipelines.yml (hunk 7 above)
Comment (SukruthGrandhiNI) [reply azure-pipelines.yml]: done the changes

**Comment by SukruthGrandhiNI** [comment]
I have fixed all the review comments.
After merging this PR, I would be working on below tasks

1. Need to create a official azure pipeline which is linked to master branch of nimi-bot- [task link](https://ni.visualstudio.com/DevCentral/_sprints/backlog/MTO%20NIB/DevCentral/21Q3/09?workitem=1631392)
2. [Security] If the PR is raised by external contributor then pipeline build should start only after approval from the owner of this repo. This is done for the sake of security- [task link](https://ni.visualstudio.com/DevCentral/_sprints/backlog/MTO%20NIB/DevCentral/21Q3/09?workitem=1631393)
3. Add  stage for Linux OS - Redhat - [task link](https://ni.visualstudio.com/DevCentral/_sprints/backlog/MTO%20NIB/DevCentral/21Q3/09?workitem=1631394)
4. Add stage for Windows 64 bit OS but execute tests on 32 bit python - [task link](https://ni.visualstudio.com/DevCentral/_sprints/backlog/MTO%20NIB/DevCentral/21Q3/09?workitem=1631405)
5. Add the pipeline build status badge to the readme file - [task link](https://ni.visualstudio.com/DevCentral/_sprints/backlog/MTO%20NIB/DevCentral/21Q3/09?workitem=1631398)

Apart from above tasks there is one more issue seen in azure pipelines.
Azure pipeline jobs are taking more time to connect to the agents - [Bug link](https://dev.azure.com/ni/DevCentral/_workitems/edit/1629292)

**Comment by sbethur** [comment]
> 
> 
> > Right now we have the ability to run on 32-bit and 64-bit Windows OSes. This is actually important because we have bit specific code to deal with different packing in either platform.
> 
> 32 windows is not in OS supported list, Please refer the link: https://dev.azure.com/ni/DevCentral/_git/Policy?path=%2Fproduct_and_technology%2FOperating-System-Support.md&version=GBmaster&_a=preview
> 
> > In this AzDO flavor it isn't clear which bitness this is running on. Can you make clear the bitness? Can you comment on the other bitness and Linux plans?
> > Since 32 bit windows is not there, so I am thinking to not to add bitness here. If you want to explicitly call out I can make the change.

Spoke with @SukruthGrandhiNI offline and clarified that we're talking about validating with 32-bit driver runtimes and not on 32-bit Windows.

**Comment by SukruthGrandhiNI** [comment]
@marcoskirsch @sbethur 

I have fixed all the review comments.

One more thing, I have reduced tox parallel threads from 3 to 2..
I would go with parallel threads 2 for now and later we can increase to 3

There are 5 agents.
4 agents for running windows 64 bit system tests.
1 agent for ensuring the codegen is update to date.


Please look at the previous comment about next set of activities will be done after this changes has submitted.

Latest pipeline build: [link](https://ni.visualstudio.com/Users/_build/results?buildId=1341944&view=results)

Thank you.

**Comment by marcoskirsch** [comment]
> One more thing, I have reduced tox parallel threads from 3 to 2..

Why?

**Comment by SukruthGrandhiNI** [comment]
> > One more thing, I have reduced tox parallel threads from 3 to 2..
> 
> Why?

Intermittent issue is seen when parallel is 4, log https://ni.visualstudio.com/8c36cb1d-ece7-4ec9-9c6b-409d081af0e8/_apis/build/builds/1341296/logs/31. This issue is seen once in 3-4 runs

Issue is not seen  when we reduce parallel from 4 to 2, tested the pipeline 6 times.

Idea here is, Initially will create robust and reliable pipeline, later we can increase efficiency like
1) Increase the agents from 4 to X in the VM
2) Increase the tox parallel runs from 2 to X.

@marcoskirsch

**Comment by SukruthGrandhiNI** [comment]
Please read my comment for the pre-context.

Increase tox parallel threads to 4. 
Filed an intermittent issue https://github.com/ni/nimi-python/issues/1658

When we decrease parallel threads to 2, issue is seen once in 8-10 runs. So decreasing the parallel threads to 2 is not solving the issue.

FYI  @sbethur @marcoskirsch

**Comment by marcoskirsch** [comment]
> Intermittent issue is seen when parallel is 4,

I don't like not knowing the root causes of intermittent issues. Do you know it? If you don't that's ok if we plan on following up later.

**Comment by SukruthGrandhiNI** [comment]
> I don't like not knowing the root causes of intermittent issues. Do you know it? If you don't that's ok if we plan on following up later.

@sbethur  has also confirmed it is a pre-existing issue. Issue is seen more often since we are recreating the tox environment for every build in the agent, 

@marcoskirsch I believe this intermittent issue is seen because of parallel thing so I will disable it. I will validate this for 30 runs. If   Intermittent issue is not seen then we can disable parallel thing now, later we can improvise the efficiency. your thoughts?

**Comment by SukruthGrandhiNI** [comment]
@sbethur @marcoskirsch 
When I remove parallel thing then 1 run is failing out of 34 runs.
Failing run log: https://ni.visualstudio.com/8c36cb1d-ece7-4ec9-9c6b-409d081af0e8/_apis/build/builds/1351782/logs/69
Total runs: https://ni.visualstudio.com/Users/_build?definitionId=7476&_a=summary

I have already created an issue to track this kind of failure: https://github.com/ni/nimi-python/issues/1658

File: azure-pipelines.yml [hunk 8]
```diff
@@ -0,0 +1,91 @@
+parameters:
+- name: module_names
+  type: object
+  default:
+  - nidigital
```
Comment (sbethur) [inline azure-pipelines.yml:5]: Add a comment about why you chose this particular order.

File: azure-pipelines.yml [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+  branches:
+    include:
+    - master
+    - releases/*
+  paths:
+    exclude:
+    - CHANGELOG.md
+    - CONTRIBUTING.md
+    - .gitattributes
+    - .gitattributes
```
Comment (sbethur) [inline azure-pipelines.yml:29]: Repeated

**Review by sbethur — APPROVED** [review]

File: azure-pipelines.yml (hunk 8 above)
Comment (SukruthGrandhiNI) [reply azure-pipelines.yml:5]: Based on time taken by the modules.
NIDCPower/nitclk are taking around 15 mins.
nifgen is taking 5 mins
dcpower, dmm, modinst, scope are taking around 2 mins.
se(Swith Executive), Switch is taking around 1 min.

Comment (marcoskirsch) [reply azure-pipelines.yml:5]: That's a brittle comments. Are the times at least consistent? Why in the world does `nidcpower` take 15 minutes?!

File: azure-pipelines.yml [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+
+# specific path
+pr:
+  branches:
+    include:
+    - master
+    - releases/*
+  paths:
+    exclude:
+    - CHANGELOG.md
```
Comment (marcoskirsch) [inline azure-pipelines.yml:26]: I don't know if this works, but if you can exclude all markdown docs `*.md` it will be more robust.

File: azure-pipelines.yml [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+    - CONTRIBUTING.md
+    - .gitattributes
+    - LICENSE
+    - VERSION
+  drafts: false
+
+stages:
+
+  - stage: ensure_code_gen_up_to_date
```
Comment (marcoskirsch) [inline azure-pipelines.yml:35]: This stage is not described in the Description of the PR.
Either describe it, or (even better) make that change separately. Smaller PRs are always better.

File: azure-pipelines.yml [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+    jobs:
+    - job: ensure_code_gen_up_to_date
+      workspace:
+        clean: all
+      steps:
+        - checkout: self  
+          clean: true
+          fetchDepth: $(clone_depth)
+        - task: PythonScript@0
+          displayName: Ensure codegen up to date
```
Comment (marcoskirsch) [inline azure-pipelines.yml:50]: Job is `ensure_code_gen_up_to_date`, but here you type "codegen".
Let's be precise and correct:

* Abbreviation is "codegen" as used throughout the repo. So no underscore between code and gen.
* Either say codegen here, or spell out "code generation"
* Up to date should be "up-to-date"
* Is this self-explanatory enough? displayName could be better: "Ensure changes to code generation are committed to repository".

File: azure-pipelines.yml [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+          inputs:
+            scriptSource: inline
+            script: |
+              import os
+              import sys
+              import subprocess
+
+              # To verify code is generated
```
Comment (marcoskirsch) [inline azure-pipelines.yml:58]: Don't paraphrase, instead match the contents of displayName.

File: azure-pipelines.yml [hunk 14]
```diff
@@ -0,0 +1,90 @@
...
+              if os.system('git config user.email "dummy@ni.com"') != 0:
+                sys.exit(-1)
```
Comment (marcoskirsch) [inline azure-pipelines.yml:60]: Lots of checks for error conditions but they all return the same code. Help developers debug this by using distinct codes and/or printing a helpful error to stderr

File: azure-pipelines.yml [hunk 15]
```diff
@@ -0,0 +1,90 @@
...
+              if os.system('git add .') != 0:
+                print("Unable to add the files", file=sys.stderr)
+                sys.exit(-1)              
+              os.system('git commit -m "test"') # return code is 1 if there are files to commit
+              commit_id_after_code_generator = os.system("git log -1 --format=%h")
+              if commit_id_before_code_generator != commit_id_after_code_generator:
+                print("Run codegen and include the generated files in the PR", file=sys.stderr)
+                sys.exit(-1)
+              print("codgen is up to date")
```
Comment (marcoskirsch) [inline azure-pipelines.yml:76]: "All changes to code generation were committed to the repository."

**Review by marcoskirsch — CHANGES_REQUESTED** [review]

File: azure-pipelines.yml (hunk 8 above)
Comment (sbethur) [reply azure-pipelines.yml:5]: I think he meant nidigital and not nidcpower takes 15 mins, which is due to `test_self_calibrate` taking a long time.

File: azure-pipelines.yml (hunk 10 above)
Comment (sbethur) [reply azure-pipelines.yml:26]: Support for wildcards was added just this week: https://docs.microsoft.com/en-us/azure/devops/release-notes/2021/sprint-192-update#support-for-wild-cards-in-path-filters.

File: azure-pipelines.yml [hunk 16]
```diff
@@ -0,0 +1,174 @@
...
+      dependsOn: ensure_code_gen_up_to_date 
+      workspace:
+        clean: all
+      steps:
+        - checkout: self  
+          clean: true
+          fetchDepth: $(clone_depth)
+        - script: | 
+            tools\system_tests.bat nidigital
+          displayName: nidigital system tests windows
```
[the remark this answers is not in the document]
Comment (sbethur) [reply azure-pipelines.yml]: Looking at the whole picture gives a better idea of how to names these..
image

What do you think of the names I've proposed in the screenshot?

[the remark this answers is not in the document]
Comment (SukruthGrandhiNI) [reply azure-pipelines.yml]: I have updated the names suggested by @sbethur 
FYI @marcoskirsch

File: azure-pipelines.yml [hunk 17]
```diff
@@ -0,0 +1,174 @@
...
+        clean: all
+      steps:
+        - checkout: self  
+          clean: true
+          fetchDepth: $(clone_depth)
+        - script: | 
+            tools\system_tests.bat nise
+          displayName: nise system tests windows
+
+    - job: niswitch_system_tests_windows
```
[the remark this answers is not in the document]
Comment (SukruthGrandhiNI) [reply azure-pipelines.yml]: I think you looked at previous commits.
Refactoring is already done.

File: azure-pipelines.yml (hunk 10 above)
Comment (SukruthGrandhiNI) [reply azure-pipelines.yml:26]: Will try this out.

File: azure-pipelines.yml (hunk 12 above)
Comment (SukruthGrandhiNI) [reply azure-pipelines.yml:50]: > * Up to date should be "up-to-date": we can't use '-' in job name. 

> * Is this self-explanatory enough? displayName could be better: "Ensure changes to code generation are committed to repository".
I have updated the display name as per your suggestion.
Previously, I was using naming conventions from other PR's e.g https://ni.visualstudio.com/DevCentral/_build/results?buildId=1352531&view=logs&j=abca7640-95f0-558f-0764-dca33dd54edb&t=f9f9b2ab-d6da-505d-8ca9-db34651b59c7

File: azure-pipelines.yml (hunk 17 above)
[the remark this answers is not in the document]
Comment (SukruthGrandhiNI) [reply azure-pipelines.yml]: I think you are looking at previous commits.
Refactoring is already done in the latest changes.

File: azure-pipelines.yml [hunk 18]
```diff
@@ -0,0 +1,174 @@
...
+      steps:
+        - checkout: self  
+          clean: true
+          fetchDepth: $(clone_depth)
+        - task: PythonScript@0
+          displayName: Ensure codegen up to date
+          inputs:
+            scriptSource: inline
+            script: |
+              import os
```
[the remark this answers is not in the document]
Comment (SukruthGrandhiNI) [reply azure-pipelines.yml:54]: I agree with @marcoskirsch

FYI @sbethur do let us know if you think otherwise?

File: azure-pipelines.yml (hunk 15 above)
Comment (SukruthGrandhiNI) [reply azure-pipelines.yml]: used print statements before exit

File: azure-pipelines.yml (hunk 18 above)
[the remark this answers is not in the document]
Comment (sbethur) [reply azure-pipelines.yml]: I like a separate file; it's better for re-usability.

File: azure-pipelines.yml (hunk 10 above)
Comment (SukruthGrandhiNI) [reply azure-pipelines.yml:26]: Will try this with next change.

File: .template_system_tests.yml [hunk 19]
```diff
@@ -0,0 +1,24 @@
+parameters:
+  module_name: ""
+  os_name: ""
+  clone_depth: 10
```
Comment (marcoskirsch) [inline .template_system_tests.yml:6]: why do we need clone depth > 1?

File: azure-pipelines.yml [hunk 20]
```diff
@@ -0,0 +1,67 @@
+parameters:
+- name: module_names
+  type: object
+  default: # Order of the modules is based on test execution time, longest firt.
```
Comment (marcoskirsch) [inline azure-pipelines.yml:6]: firt ---> first

typo

File: .template_system_tests.yml [hunk 21]
```diff
@@ -0,0 +1,24 @@
+parameters:
```
Comment (marcoskirsch) [inline .template_system_tests.yml:3]: how about 1-line comment
```
#
# Run system tests for the specified module on the specified platform.
#
```

File: .template_system_tests.yml [hunk 22]
```diff
@@ -0,0 +1,24 @@
...
+  module_name: ""
+  os_name: ""
```
Comment (marcoskirsch) [inline .template_system_tests.yml:5]: `os_name` is not a precise name, I think, because the options are linux, win32, win64 (right?)
But in reality both win32 and win64 are executed on the same OS (Windows 10 64-bit).

A more precise term would be `platform`.

File: azure-pipelines.yml [hunk 23]
```diff
@@ -0,0 +1,67 @@
+parameters:
```
Comment (marcoskirsch) [inline azure-pipelines.yml:3]: consider 1 line overview comment

```
#
# nimi-bot pipeline for running all system tests as part of Continuous Integration.
#
```

File: .template_system_tests.yml (hunk 21 above)
Comment (marcoskirsch) [reply .template_system_tests.yml:3]: and/or can we give this file any name? something that provide more context like `run_module_system_tests.yml`

File: azure-pipelines.yml (hunk 10 above)
Comment (marcoskirsch) [reply azure-pipelines.yml:28]: did you try out? or did you mean next PR?

File: azure-pipelines.yml [hunk 24]
```diff
@@ -0,0 +1,69 @@
...
+  - nidmm
+  - niscope
+  - nimodinst
+  - nise
+  - niswitch
+
+# A pipeline with no CI trigger
+trigger: none
+
+# specific path
```
Comment (marcoskirsch) [inline azure-pipelines.yml:20]: I don't think this comment is accurate, not sure where it came from. Should be removed or improved `# When should the nimi-bot CI be kicked off.`

File: azure-pipelines.yml [hunk 25]
```diff
@@ -0,0 +1,69 @@
...
+    - CHANGELOG.md
+    - CONTRIBUTING.md
+    - .gitattributes
+    - LICENSE
+    - VERSION
+  drafts: false
+
+stages:
+
+  - stage: ensure_codegen_up_to_date
```
Comment (marcoskirsch) [inline azure-pipelines.yml:37]: @sbethur - this is a good check to have in place, but seems strange to me to run this on nimi-bot as it is not a system test.
We run unit tests, linter tests, codecoverage on TravisCI and this seems closer to those. No? What is our long term plan for all that?

@SukruthGrandhiNI - this is a good check to have in place, but it seems orthogonal to this PR. Can we make this addition in its own PR?

File: azure-pipelines.yml [hunk 26]
```diff
@@ -0,0 +1,69 @@
...
+      name: nimi-bot
+      demands:
+        - nimi-bot -equals True
+        - agent.os -equals Windows_NT
+    jobs:
+    - ${{each module_name in parameters.module_names}}:
+      - template: .template_system_tests.yml
+        parameters:
+          module_name: ${{module_name}}
+          os_name: win64
```
Comment (marcoskirsch) [inline azure-pipelines.yml:69]: the fact that agent.os is "Linux" or "Windows_NT" makes me feel even more sure that `os_name` is not a good name for this parameter :)

File: tools/ensure_codegen_up_to_date.py [hunk 27]
```diff
@@ -0,0 +1,89 @@
...
+    add_codgen_files_to_git_staging_area()
+    commit_git_staging_area_files()
+
+    commit_id_after_code_generator = get_commit_id()
+
+    if commit_id_before_code_generator != commit_id_after_code_generator:
+        print("Run codegen and include the generated files in the PR", file=sys.stderr)
+        sys.exit(-1)
+    
+    print("codgen is up to date")
```
Comment (marcoskirsch) [inline tools/ensure_codegen_up_to_date.py:89]: Typo.
Consider rephrasing that is less ambiguous:

> All changes to code generation are committed to repository

**Review by marcoskirsch — CHANGES_REQUESTED** [review]

File: .template_system_tests.yml (hunk 19 above)
Comment (SukruthGrandhiNI) [reply .template_system_tests.yml:6]: Changed to 1.

File: azure-pipelines.yml (hunk 24 above)
Comment (SukruthGrandhiNI) [reply azure-pipelines.yml:20]: This pipeline will trigger only for the PR runs not after submitting the changes.

File: azure-pipelines.yml (hunk 26 above)
Comment (SukruthGrandhiNI) [reply azure-pipelines.yml:69]: As suggested changed to platform

File: azure-pipelines.yml (hunk 24 above)
Comment (marcoskirsch) [reply azure-pipelines.yml:20]: yeah but why does the comment say "specific path" ?

Comment (SukruthGrandhiNI) [reply azure-pipelines.yml:20]: I have removed the comment.

Comment (SukruthGrandhiNI) [reply azure-pipelines.yml]: Generally we provide include paths in this case we are only providing exclude paths. So I removed the comment.

**Comment by SukruthGrandhiNI** [comment]
>> did you try out? or did you mean next PR?
I mean for the next PR I will try that. 
Also, it is easy to test once this changes goes in.

@marcoskirsch

File: azure-pipelines.yml (hunk 25 above)
Comment (SukruthGrandhiNI) [reply azure-pipelines.yml:36]: @marcoskirsch I think we can have it here because we are running system tests on the generated folders. so we can validate generated folders before running the system tests.
I am also fine removing this check from this PR.

Comment (sbethur) [reply azure-pipelines.yml:36]: The long term plan is to do all the steps in a single pipeline using GitHub Actions. Till we get there I'm not too concerned about where we're doing this step, but having it somewhere is definitely a net positive. If adding to TravisCI is not trivial, then it might not be worth the time since we want to move away from it.

Either way, it would be ideal to do it in its own PR.

Comment (SukruthGrandhiNI) [reply azure-pipelines.yml]: I have removed this check from this PR

**Review by marcoskirsch — APPROVED** [review]
